cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ProgramLifecycleService.java

```diff
@@ -189,7 +201,8 @@ public ProgramStatus getProgramStatus(ProgramId programId) throws NotFoundExcept
    * @return the {@link ProgramSpecification} for the specified {@link ProgramId program}
    */
   @Nullable
-  public ProgramSpecification getProgramSpecification(ProgramId programId) {
+  public ProgramSpecification getProgramSpecification(ProgramId programId) throws Exception {
+    ensureAccess(programId);
```

Comment:
Moved this to be after fetch `appSpec`.  Just don't want this authorization checks suppressing other exceptions.